pkg/migrate/util.go

```diff
@@ -0,0 +1,342 @@
+
+	// query for api service instances by reference to a revision name
+	for _, rev := range revs {
+		wg.Add(1)
+
+		go func(revision *v1.ResourceInstance) {
+			defer wg.Done()
+
+			q := map[string]string{
+				"query": queryFunc(revision.Name),
+			}
+			url := m.Cfg.GetInstancesURL()
+
+			// Passing down apiservice name (ri.Name) for logging purposes
+			// Possible future refactor to send context through to get proper resources downstream
+			err := updateSvcInstance(url, q, revision, m)
+			errCh <- err
+		}(rev)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
+}
+
+func updateSvcInstance(
+	resourceURL string, query map[string]string, revision *v1.ResourceInstance, m *MarketplaceMigration) error {
+	resources, err := m.Client.GetAPIV1ResourceInstancesWithPageSize(query, resourceURL, 100)
+	if err != nil {
+		return err
+	}
+
+	wg := &sync.WaitGroup{}
+	errCh := make(chan error, len(resources))
+
+	for _, resource := range resources {
+		wg.Add(1)
+
+		go func(svcInstance *v1.ResourceInstance) {
+			defer wg.Done()
+
+			err := handleSvcInstance(svcInstance, revision, m)
+			if err != nil {
+				errCh <- err
+			}
+
+		}(resource)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	for e := range errCh {
+		if e != nil {
+			return e
+		}
+	}
+
+	return nil
+}
+
+func processAccessRequestDefinition(oauthScopes map[string]string, m *MarketplaceMigration) (string, error) {
+	ard, err := registerAccessRequestDefinition(oauthScopes)
+	if err != nil {
+		return "", err
+	}
+
+	newARD, err := m.Client.CreateOrUpdateResource(ard)
+	if err != nil {
+		return "", err
+	}
+	var ardRIName string
+	if newARD != nil {
+		ard, err := newARD.AsInstance()
+		if err == nil {
+			m.Cache.AddAccessRequestDefinition(ard)
+		} else {
+			return "", err
+		}
+		ardRIName = ard.Name
+	}
+	return ardRIName, nil
+}
+
+func getCredentialRequestPolicies(authPolicies []string) ([]string, error) {
+	var credentialRequestPolicies []string
+
+	for _, policy := range authPolicies {
+		if policy == apic.Apikey {
+			credentialRequestPolicies = append(credentialRequestPolicies, provisioning.APIKeyCRD)
+		}
+		if policy == apic.Oauth {
+			credentialRequestPolicies = append(credentialRequestPolicies, []string{provisioning.OAuthPublicKeyCRD, provisioning.OAuthSecretCRD}...)
+		}
+	}
+
+	return credentialRequestPolicies, nil
+
+}
+
+func checkCredentialRequestDefinitions(credentialRequestPolicies []string, m *MarketplaceMigration) []string {
+	// remove any crd not in the cache
+	knownCRDs := make([]string, 0)
+	for _, policy := range credentialRequestPolicies {
+		if def, err := m.Cache.GetCredentialRequestDefinitionByName(policy); err == nil && def != nil {
+			knownCRDs = append(knownCRDs, policy)
+		}
+	}
+
+	return knownCRDs
+}
+
+func registerAccessRequestDefinition(scopes map[string]string) (*mv1a.AccessRequestDefinition, error) {
+	oauthScopes := make([]string, 0)
+	for scope := range scopes {
```

Comment:
I don't think the scope stuff here is used anymore

Reply:
right... removing...